### PR TITLE
experiment with u-kanren

### PR DESCRIPTION
**Pull request by jitwit**
Hey! I thought it would be fun to take a stab at adding micro-kanren  (#81) .

I ended up using `assp` and `equal?` for the implementation, which I added to `rnrs.ss`. I gather that the `equal?` is improper since it doesn't do cyclic inputs, which I'm guessing is what's wanted?

Anyhow, the actual test cases I added were based on stuff in the linked paper. Are there other examples that should be there?

### CODE REVIEW

File: test/micro-kanren.ss [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+  ;;; Goals
+  (define (succeed s.c) `(,s.c))
+  (define (fail s.c) '())
+
+  (define (== u v)
+    (lambda (s.c)
+      (let ((s (unify u v (car s.c))))
+        (if s (succeed `(,s . ,(cdr s.c))) (fail s)))))
+
+  (define (run n goal)
```
Comment (eholk) [inline test/micro-kanren.ss]: Does `run` need a query var as well, like `run*`?

File: test/micro-kanren.ss [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+  (define (disj g1 g2)
+    (lambda (s.c)
+      (interleave (g1 s.c) (g2 s.c))))
+
+  (define (conj g1 g2)
+    (lambda (s.c)
+      (bind (g1 s.c) g2)))
+
+  (define (interleave $1 $2)
```
Comment (eholk) [inline test/micro-kanren.ss]: I like that you've put meaningful names here like `interleave` as opposed to `mplus`. It would be helpful to add comments that explain how the names match up with the functions in the paper though.

File: test/micro-kanren.ss (hunk 1 above)
Comment (jitwit) [reply test/micro-kanren.ss]: Ah yes, I hadn't settled on how to extract the results. In this last commit, I followed `run n var goal` and `run* var goal` from the paper where the latter forces the whole stream.

File: test/micro-kanren.ss (hunk 2 above)
Comment (jitwit) [reply test/micro-kanren.ss]: Makes sense! I think most of the naming differences are addressed now. 

I wanted to come up with a more evocative name than `bind` as well, but failed

File: test/micro-kanren.ss [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+            ((pair? v) (cons (lookup* (car v) s)
+                             (lookup* (cdr v) s)))
+            (else v))))
+
+  (define (ext-s var value s)
+    `((,var . ,value) . ,s))
+
+  (define (unify u v s)
+    (let ((u (lookup u s)) (v (lookup v s)))
+      (cond ((eq? u v) s)
```
[the remark this answers is not in the document]
Comment (jitwit) [reply test/micro-kanren.ss]: I think you're right. The tests pass just the same without it. In most of the Kanren material I've skimmed it is the first clause, though I'm not sure why.

**Comment by jitwit** [comment]
Hi @eholk, cool! I made some updates to address your comments and couldn't help but add more relations/tests, too.

 Let me know if I missed anything or whatever else should be tweaked

File: test/micro-kanren.ss [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+  (define (fail s.c) '())
+
+  (define (== u v)
+    (lambda (s.c)
+      (lambda ()
+        (let ((s (unify u v (car s.c))))
+          (if s (succeed `(,s . ,(cdr s.c))) (fail s))))))
+
+  ;;; Use int c in s.c pair to introduce a new variable to give to f
+  ;;; to build a goal.
```
Comment (amirouche) [inline test/micro-kanren.ss]: Similarly to `interleave` that replace `mplus`. It might be better to rename `s.c` to `substitutions+constant` or better (imo) `substitutions+index` (or even better move the constant / index to a stateful procedure `fresh-variable` it will allow to tremendously simplify the code and avoid the all around use of `car` to extract the substitution list).

At least, it might require a comment about the structure of `s.c` to explain how it is used in particular how the substitution list works in plain english (even if it is explained in `lookup` (and the paper)).

Comment (amirouche) [reply test/micro-kanren.ss]: ?

Comment (amirouche) [reply test/micro-kanren.ss]: WDYT?

File: test/micro-kanren.ss (hunk 2 above)
Comment (eholk) [reply test/micro-kanren.ss]: Thanks for the comments. This is much clearer now!

File: test/micro-kanren.ss [hunk 5]
```diff
@@ -142,22 +277,46 @@
...
-                                    (appendo vx vy '(0 1 2)))))))))
+                   (run* vx (conj (element-of vx '(1 2 3))
+                                  (== vx 0))))
+           (equal? '((() (s c h i s m))
+                     ((s) (c h i s m))
+                     ((s c) (h i s m))
+                     ((s c h) (i s m))
+                     ((s c h i) (s m))
+                     ((s c h i s) (m))
+                     ((s c h i s m) ()))
```
Comment (eholk) [inline test/micro-kanren.ss]: Nice test case! :smile:

**Comment by eholk** [comment]
I'll go ahead and merge this PR. Thank you so much for the contribution!

Feel free to submit another PR if you want to incorporate any of @amirouche's suggestions.

File: test/micro-kanren.ss (hunk 4 above)
Comment (eholk) [reply test/micro-kanren.ss]: Would the stateful procedure essentially be doing a `gensym`? I'd be a little concerned about adding too much state, because some of the elegance of uKanren is it's purely functional manner. Although, using state under the covers would make sense if we wanted to build this up to a full scale logic programming system.

I think adding `substitution+constant` in a comment wouldn't hurt. As far as names go though, I don't mind the short one. Out of curiosity, @jitwit, was there a problem parsing `s/c`, which I think is what the paper used? If so, we probably have a reader bug.

Comment (jitwit) [reply test/micro-kanren.ss]: No issues with the reader! I had just found the `s/c` name slightly confusing at first. Something broader like `state` might make more sense in case the representation changes or expands down the line.
